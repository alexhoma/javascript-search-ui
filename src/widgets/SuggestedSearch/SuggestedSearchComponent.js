/**
 * @jsx h
 */

import { h, Component } from 'preact';
import {
    highlightSuggestion,
    selectNextSuggestion,
    selectPreviousSuggestion,
    selectActiveSuggestion
} from './helpers';

/**
 * Actions
 */
import {
    simpleSearchAction,
    suggestedSearchAction
} from "./suggestedSearchActions";
import Template from "../Template";

/**
 * Suggested Search Component
 */
class SuggestedSearchComponent extends Component {
    constructor() {
        super();
        this.state = {
            q: '',
            currentSuggestions: []
        };
    }

    componentWillReceiveProps(props) {
        /**
         * Check suggestions available
         * if some, prepend the current query to the other suggestions array
         * else, only append the current query to the suggestions array
         */
        const suggests = (props.data && props.data.suggests)
            ? [this.state.q, ...props.data.suggests]
            : [this.state.q]
        ;

        /**
         * Prepare suggestions array
         */
        this.setState({
            currentSuggestions: suggests.map((suggest, key) => {
                return {
                    isActive: (0 === key),
                    name: suggest,
                    htmlName: highlightSuggestion(this.state.q, suggest)
                }
            })
        });
    }

    handleSearch = (e) => {
        const {
            startSearchOn,
            environmentId,
            currentQuery,
            client
        } = this.props;

        /**
         * Set the current query text
         */
        this.setState({q: e.target.value});

        /**
         * Search when string is bigger than {startSearchOn}
         */
        if (e.target.value.length < startSearchOn) {
            this.setState({currentSuggestions: []});
            return;
        }

        /**
         * Dispatch suggested search action
         */
        suggestedSearchAction(
            {
                queryText: e.target.value
            },
            {
                environmentId,
                currentQuery,
                client
            }
        )
    };

    handleSuggestionsNavigation = (e) => {
        /**
         * When user hits arrow down
         */
        if (e.code === 'ArrowDown') {
            this.setState({
                currentSuggestions: selectNextSuggestion(
                    this.state.currentSuggestions
                ),
                q: selectActiveSuggestion(
                    this.state.currentSuggestions
                )
            });
        }

        /**
         * When user hits arrow up
         */
        if (e.code === 'ArrowUp') {
            /**
             * Prevent cursor to go at the starting point of the line
             */
            e.preventDefault();

            this.setState({
                currentSuggestions: selectPreviousSuggestion(
                    this.state.currentSuggestions
                ),
                q: selectActiveSuggestion(
                    this.state.currentSuggestions
                )
            });
        }

        /**
         * When user hits enter
         */
        if (e.code === 'Enter') {
            this.setState({
                q: selectActiveSuggestion(
                    this.state.currentSuggestions
                ),
                currentSuggestions: []
            });

            const {
                environmentId,
                currentQuery,
                client
            } = this.props;

            simpleSearchAction(
                {
                    queryText: this.state.q
                },
                {
                    environmentId,
                    currentQuery,
                    client
                }
            )
        }
    };

    handleSuggestionClick = (e) => {
        this.setState({
            q: e.target.innerText,
            currentSuggestions: []
        });

        const {
            environmentId,
            currentQuery,
            client
        } = this.props;

        simpleSearchAction(
            {
                queryText: e.target.innerText
            },
            {
                environmentId,
                currentQuery,
                client
            }
        )
    };

    handleSearchInputFocusedOut = (e) => {
        /**
         * It handles when a user focuses out the search input
         * If is not clicking on the suggestions box
         * The suggestions are cleared and panel closes
         */
        if (
            null === e.relatedTarget ||
            false === e.relatedTarget.classList.contains('as-suggestedSearch__box')
        ) {
            this.setState({currentSuggestions: []});
        }

        return false;
    };

    clearSearch = () => {
        const {
            environmentId,
            currentQuery,
            client
        } = this.props;

        simpleSearchAction(
            {
                queryText: ''
            },
            {
                environmentId,
                currentQuery,
                client
            }
        )
    };

    render() {
        const {
            dirty,
            placeholder,
            autofocus,
            classNames: {
                container: containerClassName,
                input: inputClassName,
                clearSearch: clearSearchClassName,
                box: boxClassName,
                suggestion: suggestionClassName,
                activeSuggestion: activeSuggestionClassName
            },
            template: {
                clearSearch: clearSearchTemplate
            },
            currentQuery: {
                q: currentQueryText
            }
        } = this.props;

        const {
            currentSuggestions
        } = this.state;

        return (
            <div className={`as-suggestedSearch ${containerClassName}`}>
                <input
                    type={`text`}
                    value={currentQueryText}
                    className={`as-suggestedSearch__input ${inputClassName}`}
                    placeholder={placeholder}
                    autofocus={autofocus}

                    onInput={this.handleSearch}
                    onKeyDown={this.handleSuggestionsNavigation}
                    onBlur={this.handleSearchInputFocusedOut}
                />

                {(currentQueryText.length !== 0)
                    ? (
                        <div
                            className={`as-simpleSearch__clearSearch ${clearSearchClassName}`}
                            onClick={this.clearSearch}
                        >
                            <Template template={clearSearchTemplate}/>
                        </div>
                    ) : null
                }

                <div
                    tabIndex={`0`}
                    className={`as-suggestedSearch__box ${boxClassName}`}
                    style={{
                        display: currentSuggestions.length > 1 && !dirty
                            ? 'block'
                            : 'none'
                    }}
                >
                    {currentSuggestions.map((suggestion, key) => {
                        return (0 !== key)
                            ? <div
                                className={
                                    `as-suggestedSearch__suggestion ` +
                                    `${suggestionClassName} ` +
                                    `${suggestion.isActive
                                        ? activeSuggestionClassName
                                        : ''
                                    }`
                                }
                                dangerouslySetInnerHTML={{
                                    __html: suggestion.htmlName
                                }}
                                onClick={this.handleSuggestionClick}
                            />
                            : null
                    })}
                </div>
            </div>
        )
    }
}

SuggestedSearchComponent.defaultProps = {
    placeholder: '',
    autofocus: false,
    startSearchOn: 0,
    classNames: {
        container: '',
        input: '',
        clearSearch: '',
        box: '',
        suggestion: '',
        activeSuggestion: 'as-suggestedSearch__suggestion--active'
    },
    template: {
        clearSearch: 'x'
    }
};

export default SuggestedSearchComponent;