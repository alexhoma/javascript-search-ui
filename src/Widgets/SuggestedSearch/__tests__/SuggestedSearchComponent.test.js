/** @jsx h */
import {h} from 'preact';
import {shallow} from 'preact-render-spy';
import SuggestedSearchComponent from './../SuggestedSearchComponent';

describe('<SuggestedSearchComponent />',() => {
    const defaultProps = {
        target: '#some-tagret'
    };
    const receivedData = {
        suggests: [
            'suggestion 1',
            'suggestion 2',
            'suggestion 3'
        ]
    };

    it('should render with default props', () => {
        const tree = shallow(
            <SuggestedSearchComponent {...defaultProps} />
        );

        /**
         * Testing the default props tree
         */
        expect(tree.component().props).toMatchSnapshot();

        /**
         * Testing rendered html
         */
        expect(tree).toMatchSnapshot();
    });
    it('should render with custom props', () => {
        /**
         * Full custom props available object
         */
        const customProps = {
            placeholder: 'Type to get suggestions...',
            autofocus: true,
            classNames: {
                container: 'custom-container-class',
                input: 'custom-input-class',
                box: 'custom-box-class',
                suggestion: 'custom-suggestion-class',
                activeSuggestion: 'custom-active-suggestion-class'
            }
        };

        const tree = shallow(
            <SuggestedSearchComponent
                {...defaultProps}
                {...customProps}
            />
        );

        /**
         * Testing the default props tree
         */
        expect(tree.component().props).toMatchSnapshot();

        /**
         * Testing rendered html
         */
        expect(tree).toMatchSnapshot();
    });
    it('should handleSearch() on input event', () => {
        const tree = shallow(
            <SuggestedSearchComponent {...defaultProps} />
        );

        /**
         * HandleSearch() mock
         */
        tree.find('input')[0].attributes.onInput = jest.fn();
        tree.find('input').simulate('input', 'i want to get suggestions');

        expect(
            tree.find('input').attr('onInput')
        ).toBeCalledWith('i want to get suggestions');
    });
    it('should componentWillReceiveProps() with suggestions and render it', () => {
        const tree = getUpdatedSuggestedSearchComponent();

        /**
         * Check suggestions state
         */
        expect(tree.state('currentSuggestions')).toMatchSnapshot();
        /**
         * Check rendered html
         */
        expect(tree).toMatchSnapshot();
    });
    it('should handleSuggestionsNavigation() and re-render the list being the last suggestion the active one', () => {
        const tree = getUpdatedSuggestedSearchComponent();
        const mockedArrowDownEvent = { code: 'ArrowDown' };

        /**
         * Navigate through suggestions
         * 3 suggestions down
         */
        tree.find('input').simulate('keyDown', mockedArrowDownEvent);
        tree.find('input').simulate('keyDown', mockedArrowDownEvent);
        tree.find('input').simulate('keyDown', mockedArrowDownEvent);

        /**
         * This last event won't do anything
         * since there are only three suggestions
         * to navigate
         */
        tree.find('input').simulate('keyDown', mockedArrowDownEvent);

        expect(tree.state('currentSuggestions')).toMatchSnapshot();
        expect(tree).toMatchSnapshot();
    });
    it('should handleSuggestionsNavigation() and re-render the list being the first suggestion the active one', () => {
        const tree = getUpdatedSuggestedSearchComponent();
        const mockedArrowDownEvent = { code: 'ArrowDown' };
        const mockedArrowUpEvent = { code: 'ArrowUp', preventDefault: jest.fn() };

        /**
         * Navigate through suggestions
         * 2 suggestions down, 1 suggestion up
         */
        tree.find('input').simulate('keyDown', mockedArrowDownEvent);
        tree.find('input').simulate('keyDown', mockedArrowDownEvent);
        tree.find('input').simulate('keyDown', mockedArrowUpEvent);

        expect(tree.state('currentSuggestions')).toMatchSnapshot();
        expect(tree).toMatchSnapshot();
    });
    it('should handleSuggestionClick(), select the clicked suggestion and close the box', () => {
        const tree = getUpdatedSuggestedSearchComponent();
        const mock = jest.fn();

        tree.find('.asui-suggestedSearch--suggestion')[2].attributes.onClick = mock;
        tree.find('.asui-suggestedSearch--suggestion').at(2).simulate('click');
        tree.setState({
            q: tree.find('.asui-suggestedSearch--suggestion').at(2).text(),
            currentSuggestions: []
        });

        expect(mock).toBeCalled();
        expect(tree).toMatchSnapshot();
    });
    it('should handleSearchInputFocusedOut() and close the suggestions box', () => {
        const tree = getUpdatedSuggestedSearchComponent();
        const mockedBlurEvent = { relatedTarget: null };

        /**
         * Click outside input value,
         * and outside relatedTarget value "suggestions box"
         */
        tree.find('input').simulate('blur', mockedBlurEvent);

        expect(tree).toMatchSnapshot();
    });

    /**
     * Return a Shallowed Suggested search component
     * with an updated state and props
     */
    const getUpdatedSuggestedSearchComponent = () => {
        const tree = shallow(
            <SuggestedSearchComponent {...defaultProps} />
        );

        /**
         * Simulation of handleSearch functionality
         * setting the state of the query
         * and changing its props
         */
        tree.setState({q: 'i want to get suggestions'});

        /**
         * Triggering componentWillReceiveProps(...props)
         * with this preact-render-spy hack
         */
        tree.render(
            <SuggestedSearchComponent
                {...defaultProps}
                data={receivedData}
            />
        );

        return tree;
    }
});