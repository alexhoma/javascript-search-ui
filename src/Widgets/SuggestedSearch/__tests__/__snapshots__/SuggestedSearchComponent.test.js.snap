// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SuggestedSearchComponent /> should componentWillReceiveProps() with suggestions and render it 1`] = `
Array [
  Object {
    "htmlName": "<em>i&nbsp;want&nbsp;to&nbsp;get&nbsp;suggestions</em>",
    "isActive": true,
    "name": "i want to get suggestions",
  },
  Object {
    "htmlName": "suggestion&nbsp;1",
    "isActive": false,
    "name": "suggestion 1",
  },
  Object {
    "htmlName": "suggestion&nbsp;2",
    "isActive": false,
    "name": "suggestion 2",
  },
  Object {
    "htmlName": "suggestion&nbsp;3",
    "isActive": false,
    "name": "suggestion 3",
  },
]
`;

exports[`<SuggestedSearchComponent /> should componentWillReceiveProps() with suggestions and render it 2`] = `
preact-render-spy (1 nodes)
-------
<div class="asui-suggestedSearch ">
  <input
    type="text"
    value="i want to get suggestions"
    class="asui-suggestedSearch--input "
    placeholder=""
    autofocus={false}
    onInput={[Function anonymous]}
    onKeyDown={[Function anonymous]}
    onBlur={[Function anonymous]}
   />
  <div
    tabIndex="0"
    class="asui-suggestedSearch--box "
    style="display: block;"
  >
    <div
      class="asui-suggestedSearch--suggestion  "
      onClick={[Function anonymous]}
    >suggestion&nbsp;1
    </div>
    <div
      class="asui-suggestedSearch--suggestion  "
      onClick={[Function anonymous]}
    >suggestion&nbsp;2
    </div>
    <div
      class="asui-suggestedSearch--suggestion  "
      onClick={[Function anonymous]}
    >suggestion&nbsp;3
    </div>
  </div>
</div>

`;

exports[`<SuggestedSearchComponent /> should handleSearchInputFocusedOut() and close the suggestions box 1`] = `
preact-render-spy (1 nodes)
-------
<div class="asui-suggestedSearch ">
  <input
    type="text"
    value="i want to get suggestions"
    class="asui-suggestedSearch--input "
    placeholder=""
    autofocus={false}
    onInput={[Function anonymous]}
    onKeyDown={[Function anonymous]}
    onBlur={[Function anonymous]}
   />
  <div
    tabIndex="0"
    class="asui-suggestedSearch--box "
    style="display: none;"
  >
  </div>
</div>

`;

exports[`<SuggestedSearchComponent /> should handleSuggestionClick(), select the clicked suggestion and close the box 1`] = `
preact-render-spy (1 nodes)
-------
<div class="asui-suggestedSearch ">
  <input
    type="text"
    value=""
    class="asui-suggestedSearch--input "
    placeholder=""
    autofocus={false}
    onInput={[Function anonymous]}
    onKeyDown={[Function anonymous]}
    onBlur={[Function anonymous]}
   />
  <div
    tabIndex="0"
    class="asui-suggestedSearch--box "
    style="display: none;"
  >
  </div>
</div>

`;

exports[`<SuggestedSearchComponent /> should handleSuggestionsNavigation() and re-render the list being the first suggestion the active one 1`] = `
Array [
  Object {
    "htmlName": "<em>i&nbsp;want&nbsp;to&nbsp;get&nbsp;suggestions</em>",
    "isActive": false,
    "name": "i want to get suggestions",
  },
  Object {
    "htmlName": "suggestion&nbsp;1",
    "isActive": true,
    "name": "suggestion 1",
  },
  Object {
    "htmlName": "suggestion&nbsp;2",
    "isActive": false,
    "name": "suggestion 2",
  },
  Object {
    "htmlName": "suggestion&nbsp;3",
    "isActive": false,
    "name": "suggestion 3",
  },
]
`;

exports[`<SuggestedSearchComponent /> should handleSuggestionsNavigation() and re-render the list being the first suggestion the active one 2`] = `
preact-render-spy (1 nodes)
-------
<div class="asui-suggestedSearch ">
  <input
    type="text"
    value="suggestion 1"
    class="asui-suggestedSearch--input "
    placeholder=""
    autofocus={false}
    onInput={[Function anonymous]}
    onKeyDown={[Function anonymous]}
    onBlur={[Function anonymous]}
   />
  <div
    tabIndex="0"
    class="asui-suggestedSearch--box "
    style="display: block;"
  >
    <div
      class="asui-suggestedSearch--suggestion  asui-active"
      onClick={[Function anonymous]}
    >suggestion&nbsp;1
    </div>
    <div
      class="asui-suggestedSearch--suggestion  "
      onClick={[Function anonymous]}
    >suggestion&nbsp;2
    </div>
    <div
      class="asui-suggestedSearch--suggestion  "
      onClick={[Function anonymous]}
    >suggestion&nbsp;3
    </div>
  </div>
</div>

`;

exports[`<SuggestedSearchComponent /> should handleSuggestionsNavigation() and re-render the list being the last suggestion the active one 1`] = `
Array [
  Object {
    "htmlName": "<em>i&nbsp;want&nbsp;to&nbsp;get&nbsp;suggestions</em>",
    "isActive": false,
    "name": "i want to get suggestions",
  },
  Object {
    "htmlName": "suggestion&nbsp;1",
    "isActive": false,
    "name": "suggestion 1",
  },
  Object {
    "htmlName": "suggestion&nbsp;2",
    "isActive": false,
    "name": "suggestion 2",
  },
  Object {
    "htmlName": "suggestion&nbsp;3",
    "isActive": true,
    "name": "suggestion 3",
  },
]
`;

exports[`<SuggestedSearchComponent /> should handleSuggestionsNavigation() and re-render the list being the last suggestion the active one 2`] = `
preact-render-spy (1 nodes)
-------
<div class="asui-suggestedSearch ">
  <input
    type="text"
    value="suggestion 3"
    class="asui-suggestedSearch--input "
    placeholder=""
    autofocus={false}
    onInput={[Function anonymous]}
    onKeyDown={[Function anonymous]}
    onBlur={[Function anonymous]}
   />
  <div
    tabIndex="0"
    class="asui-suggestedSearch--box "
    style="display: block;"
  >
    <div
      class="asui-suggestedSearch--suggestion  "
      onClick={[Function anonymous]}
    >suggestion&nbsp;1
    </div>
    <div
      class="asui-suggestedSearch--suggestion  "
      onClick={[Function anonymous]}
    >suggestion&nbsp;2
    </div>
    <div
      class="asui-suggestedSearch--suggestion  asui-active"
      onClick={[Function anonymous]}
    >suggestion&nbsp;3
    </div>
  </div>
</div>

`;

exports[`<SuggestedSearchComponent /> should render with custom props 1`] = `
Object {
  "autofocus": true,
  "children": Array [],
  "classNames": Object {
    "activeSuggestion": "custom-active-suggestion-class",
    "box": "custom-box-class",
    "container": "custom-container-class",
    "input": "custom-input-class",
    "suggestion": "custom-suggestion-class",
  },
  "placeholder": "Type to get suggestions...",
  "startSearchOn": 0,
  "target": "#some-tagret",
}
`;

exports[`<SuggestedSearchComponent /> should render with custom props 2`] = `
preact-render-spy (1 nodes)
-------
<div class="asui-suggestedSearch custom-container-class">
  <input
    type="text"
    value=""
    class="asui-suggestedSearch--input custom-input-class"
    placeholder="Type to get suggestions..."
    autofocus={true}
    onInput={[Function anonymous]}
    onKeyDown={[Function anonymous]}
    onBlur={[Function anonymous]}
   />
  <div
    tabIndex="0"
    class="asui-suggestedSearch--box custom-box-class"
    style="display: none;"
  >
  </div>
</div>

`;

exports[`<SuggestedSearchComponent /> should render with default props 1`] = `
Object {
  "autofocus": false,
  "children": Array [],
  "classNames": Object {
    "activeSuggestion": "asui-active",
    "box": "",
    "container": "",
    "input": "",
    "suggestion": "",
  },
  "placeholder": "",
  "startSearchOn": 0,
  "target": "#some-tagret",
}
`;

exports[`<SuggestedSearchComponent /> should render with default props 2`] = `
preact-render-spy (1 nodes)
-------
<div class="asui-suggestedSearch ">
  <input
    type="text"
    value=""
    class="asui-suggestedSearch--input "
    placeholder=""
    autofocus={false}
    onInput={[Function anonymous]}
    onKeyDown={[Function anonymous]}
    onBlur={[Function anonymous]}
   />
  <div
    tabIndex="0"
    class="asui-suggestedSearch--box "
    style="display: none;"
  >
  </div>
</div>

`;
